### DISCUSSION TITLE
Remark changes content on its own

### DISCUSSION THREAD

**Discussion by kangruixiang** [remarkjs/remark#524] [Q&A] [answered]
Hi I found that even without doing any interventions, remark changes things on its own. For example, I have following markdown content:

```
# Hello, world!

_Emphasis_ and **importance**.

---

[[test note]]

![hello](https://test.com/image.png).
```

Using this as my code:

```
const remark = require("remark");

let fileData = fs.readFileSync(file, "utf-8");
let content = remark().process(fileData);
fs.writeFileSync(file, content);
```

Here's the output: 

```
# Hello, world!

_Emphasis_ and **importance**.

* * *

\[[test note]]

![hello](https://test.com/image.png).
```

You can see that it changed the --- to * * * and added a backslash to [[test note]]. How can I stop it from doing that?

**Comment by wooorm** [comment, marked as answer]
Yup, that’s what remark does! It’s an abstract syntax tree, meaning non-essential information is lost. There’s no way around it.
There are [some options](https://github.com/remarkjs/remark/tree/main/packages/remark-stringify) to define the output style though.

**Reply by kangruixiang** [reply]
I understand the horizontal rule, but I could not for the life of me figure out why adding a backslash to [[links like this]]. Is there an option to define the output for that?

**Reply by wooorm** [reply]
There is no option for that.

The reason for that is that unescaped braces can turn into links, often unintentionally. There are other cases too, such as when you pass `a * b` through remark, it’ll turn into `a \* b`

**Reply by kangruixiang** [reply]
So how do people usually navigate around that, assuming they don't want their markdown files to have \[[links]] everywhere? Do they use remark for the processing and then regex out the backslash, or is there a more elegant way to approach that problem?

**Reply by wooorm** [reply]
They don’t view escapes as a problem, I guess?

*You* have `[[link]]` ([[link]]), remark does not change how that is rendered: `\[[link]]` is seen the same: \[[link]]

**Reply by kangruixiang** [reply]
Oh ok I see what you are saying now. It seems the problem is that the editors I use (Typora and Obsidian) already process the backslash in some way. 

I can have both `\[[test]]` and `[[test]]` in editor mode but both would come out without backslash in rendering mode. 

One last question, I'm trying to add output settings, but this isn't really changing the output:

```
let content = remark()
      .data("settings", {
        commonmark: true,
        emphasis: "*",
        strong: "*",
        rule: "-",
      })
      .process(fileData);
```

**Reply by wooorm** [reply]
You're looking for the sync version of process: https://github.com/unifiedjs/unified#processorprocesssyncfilevalue

**Reply by kangruixiang** [reply]
I'm sorry. I just could not get this right. I've been trying for so long. 

This doens't work at all. I get `content.processSync is not a function` error. 
```
let content = remark();
let result = content.processSync(fileData);
```

So I switched to unified(). At first, I do markdown to mdast to html then back to markdown:

```
var markdown = require("remark-parse");
var remark2rehype = require("remark-rehype");
var rehype2remark = require("rehype-remark");
var html = require("rehype-stringify");
var stringify = require("remark-stringify");

var processor = unified()
  .use(markdown)
  .use(remark2rehype)
  .use(rehype2remark)
  .use(stringify);

console.log(processor.processSync("# Hello world!").toString());
```

I get `Cannot set property 'Compiler' of undefined` error. But if I just do markdown and stringify, which theoretically is converting to mdast and back, it still says the same error. 

How do I manipulate markdown and still be able to set options for different markdown rules? 

Thanks a lot for all the help.

**Reply by wooorm** [reply]
What do you want to do? Format markdown? Turn markdown into HTML? Something else?

**Reply by kangruixiang** [reply]
Format markdown. I went to [here](https://github.com/remarkjs/remark/tree/main/packages/remark#use) and copied this code without changing anything:
 
```
var remark = require('remark')

remark()
  .data('settings', {commonmark: true, emphasis: '*', strong: '*'})
  .process('_Emphasis_ and __importance__', function(err, file) {
    if (err) throw err
    console.log(String(file))
  })
```

It outputted `_Emphasis_ and **importance**`. Doesn't matter what I change under setting, the output is the same.

**Reply by wooorm** [reply]
That works perfectly for me? https://runkit.com/embed/mlf8151x5iwy

**Reply by wooorm** [reply]
Your earlier remark-only version also works: https://runkit.com/embed/odjnonhplp8w, although you still needed to call `.toString()`.

For the unified version, the two plugins in the middle aren’t needed: https://runkit.com/embed/wnlc43gn1nlz

**Reply by kangruixiang** [reply]
Oh shoot that worked for me too. I reinstalled remark, and now it works. So much frustration over nothing. Thanks a lot. That was really helpful.

**Reply by wooorm** [reply]
thanks for using unified, good luck!

**Reply by kangruixiang** [reply]
Just for future references, I solved the backslash with [[links]] with [remark-wiki-link](https://github.com/landakram/remark-wiki-link#readme)
